sharktank/sharktank/types/tensors.py

```diff
@@ -1157,22 +1186,55 @@ def create(
             if "shard_dim" in extra_properties
             else None
         )
-        ts = []
-        for i in range(shard_count):
-            t_name = str(i)
-            try:
-                t = raw_tensors[t_name]
-                ts.append(t)
-                # TODO: this should be changed to tracked device affinity
-                DeviceTensorTrait(i).set(t)
-            except KeyError as e:
-                raise IOError(
-                    f"Missing component tensor '{t_name}' in {raw_tensors.keys()}"
-                ) from e
+        shard_type = extra_properties.get(
```

Comment:
Seeing the scale of this change all hardcoded feels wrong. You should be creating / expanding general functionality not hardcoding behavior through.

Look at how tensors are added / embedding. Find encapsulations for parts. There should be grouped together concepts (e.g. add_tensor should know how to handle subtypes). Same with assigning the device tensor traits. The realize is that a tensor should be able to have its placement set with each type handling it individually.

Overall I recommend enumerating the behaviors you need and land independently instead of straight shotting a single overarching implementation.